docs/debugging_guide.md

```diff
@@ -0,0 +1,362 @@
+==========================
+{: .primer-spec-toc-ignore }
+
+There are two fundamental components to debugging:
+1. Form hypotheses about the specific cause of a bug
+2. Use debugging tools and strategies to investigate those hypotheses
+
+This guide focuses on effective strategies for point 2 above, in particular through the use of a **debugger**, which allows you to pause your program's execution at key points and inspect the state of objects in memory.
+
+We also show several real examples drawn from course projects in EECS 280. In some of the cases, we've blurred out parts of screenshots where we ran the debugger on our instructor solution.
+
+TODO: link back to relevant parts of IDE or setup tutorials for debuggers
+
+## Inspect Program State
+
+A fundamental part of debugging is simply getting more information about what exactly your program is doing.
+
+### Local Variables
+
+<div class="primer-spec-callout info icon-info" markdown="1">
+Use the **variables panel** to inspect the values of variables while your debugger is paused.
+</div>
+
+Let's say I'm working on the Euchre project - there's a lot of information to keep track of during the card game. I'd like to check that everything seems in order after the cards are dealt and the players have finished making.
+
+When the debugger pauses at my breakpoint, we can see the value of local variables:
+
+<img src="images/debug_feature_variables_0.png" width="700px" />
+
+You can see some variables directly. For example, `trump` is currently `"Hearts"`.
+
+Variables may contain **"memory junk"**. `i` is currently `-11632`. That's expected in this case, because we just haven't run the line `int i=0;` quite yet. In other cases, bogus values might indicate a bug!
+
+You can **click to expand** more complex variables like pointers, arrays, or classes to inspect their contents. Below, I've expanded the `leader` variable to see the `Player` it is pointing to, then also expanded the `cards` array to take a look at their current hand.
+
+<img src="images/debug_feature_variables_1.png" width="700px" />
+
+<div class="primer-spec-callout warning" markdown="1">
+
+**Heads up!** You may need to configure your debugger to ensure full information is shown for polymorphic objects through pointers. In the example above, the `leader` variable is a `Player*` pointing to a `SimplePlayer` object. By default, the debugger may only show the base class `Player` data.
+
+To fix this in VS Code, run `-exec set print object on` from the debug console, or add the following to the `setupCommands` in your `launch.json` configuration:
+
+```json
+  "setupCommands": [
+      ...
+      {
+        "description": "Print polymorphic objects according to dynamic type",
+        "text": "set print object on"
+      }
+      ...
+  ]
+```
+{: data-variant="legacy" }
+</div>
+
+In **member functions**, you can also open up the `this` pointer. In our Euchre example, we'd see member variables from the current `Game` class instance:
```

Comment:
Might be worth a subheading here, possibly hidden from the ToC.  I can image wanting to link to this specific subsection from somewhere in the P3 (or later) spec.

Reply:
Added in dd7f165